### PR TITLE
feat: use local artifacts to fetch keys for the key-funder

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

Motivation - supports maintaining the source of truth for account keys in version control

- key-funder now fetches relayer,kathy keys from local artifacts instead of gcp secrets 
- local artifacts schema located here: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3159
- changing type of key from `BaseCloudAgentKey` to `BaseAgentKey` 

### Drive-by changes

- restricting rolesToFund to `FundableRole` type which is just Kathy,Relayer which prevent the person running the script locally to accidentally fund validator keys.

### Related issues

- fixes https://github.com/hyperlane-xyz/issues/issues/888

### Backward compatibility

Yes

### Testing

Manual

### CODE REVIEW

File: typescript/infra/src/utils/utils.ts [hunk 1]
```diff
@@ -213,6 +213,14 @@ export function assertRole(roleStr: string) {
   return role;
 }
 
+export function assertFundableRole(roleStr: string): FundableRole {
```
Comment (yorhodes) [inline typescript/infra/src/utils/utils.ts:216]: maybe enum?

File: typescript/infra/src/roles.ts [hunk 2]
```diff
@@ -6,6 +6,8 @@ export enum Role {
   Kathy = 'kathy',
 }
 
+export type FundableRole = Role.Relayer | Role.Kathy;
```
Comment (yorhodes) [inline typescript/infra/src/roles.ts:9]: ditto, enum?

File: typescript/infra/src/agents/key-utils.ts [hunk 3]
```diff
@@ -444,20 +444,18 @@ export async function persistValidatorAddressesToLocalArtifacts(
...
-  debugLog(
-    `Fetching addresses from GCP for ${context} context in ${environment} environment`,
-  );
-  return addresses[environment][context];
+export function fetchLocalKeyAddresses(role: Role): LocalRoleAddresses {
+  try {
+    // Resolve the relative path
+    const filePath = path.resolve(__dirname, `../../config/${role}.json`);
+    const data = fs.readFileSync(filePath, 'utf8');
+    const addresses: LocalRoleAddresses = JSON.parse(data);
```
Comment (yorhodes) [inline typescript/infra/src/agents/key-utils.ts:452]: why not use the readJson utility?

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 4]
```diff
@@ -835,13 +861,13 @@ async function getAddressInfo(
...
 async function getKeyInfo(
-  key: BaseCloudAgentKey,
+  key: BaseAgentKey,
   chain: ChainName,
   provider: ethers.providers.Provider,
 ) {
   return {
     ...(await getAddressInfo(key.address, chain, provider)),
-    context: key.context,
+    context: Contexts.Hyperlane, // TODO: fix
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:870]: remove TODO?

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 5]
```diff
@@ -457,6 +480,9 @@ class ContextFunder {
     const promises = chainKeyEntries.map(async ([chain, keys]) => {
       let failureOccurred = false;
       if (keys.length > 0) {
+        for (const key of keys) {
+          console.log('chain', chain, 'key', key.role, 'address', key.address);
+        }
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:485]: do we want this here?

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 6]
```diff
@@ -417,29 +422,47 @@ class ContextFunder {
...
+        roleKeysPerChain[chain][role] = [
+          new LocalAgentKey(
+            environment,
+            role,
+            fundableRoleKeys[role as FundableRole],
+            chain,
+          ),
+        ];
+      }
+    }
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:465]: this is slightly hard to read
consider combining these loops?

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts [hunk 7]
```diff
@@ -417,29 +422,47 @@ class ContextFunder {
     );
   }
 
-  // The keys here are not ReadOnlyCloudAgentKeys, instead they are AgentGCPKey or AgentAWSKeys,
-  // which require credentials to fetch. If you want to avoid requiring credentials, use
-  // fromSerializedAddressFile instead.
-  static async fromContext(
+  // the keys are retrived from the local artifacts in the infra/config/relayer.json or infra/config/kathy.json
```
Comment (yorhodes) [inline typescript/infra/scripts/funding/fund-keys-from-deployer.ts:425]: nit
```suggestion
  // the keys are retrieved from the local artifacts in the infra/config/relayer.json or infra/config/kathy.json
```

File: typescript/infra/src/utils/utils.ts (hunk 1 above)
Comment (aroralanuk) [reply typescript/infra/src/utils/utils.ts:216]: like 
```
export enum FundableRole {
  Relayer = 'Relayer',
  Kathy = 'Kathy',
}```

doesn't make too much of a difference

File: typescript/infra/src/roles.ts (hunk 2 above)
Comment (aroralanuk) [reply typescript/infra/src/roles.ts:9]: same as above

Comment (aroralanuk) [reply typescript/infra/src/roles.ts:9]: what's your rationale behind it?

File: typescript/infra/src/agents/key-utils.ts (hunk 3 above)
Comment (aroralanuk) [reply typescript/infra/src/agents/key-utils.ts:452]: done

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts (hunk 4 above)
Comment (aroralanuk) [reply typescript/infra/scripts/funding/fund-keys-from-deployer.ts:870]: done

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts (hunk 5 above)
Comment (aroralanuk) [reply typescript/infra/scripts/funding/fund-keys-from-deployer.ts:485]: done

File: typescript/infra/scripts/funding/fund-keys-from-deployer.ts (hunk 6 above)
Comment (aroralanuk) [reply typescript/infra/scripts/funding/fund-keys-from-deployer.ts:465]: done

File: typescript/infra/src/utils/utils.ts (hunk 1 above)
Comment (paulbalaji) [reply typescript/infra/src/utils/utils.ts:216]: not a fan of the enum way, just adds another place to update things in the future if needed

imo makes more sense the way it is, re-using existing Roles enum wherever poss

File: typescript/infra/src/roles.ts (hunk 2 above)
Comment (yorhodes) [reply typescript/infra/src/roles.ts:9]: https://www.becomebetterprogrammer.com/typescript-union-types-vs-enums/
enums provide a better compile time devx but its not a big deal

File: typescript/infra/src/utils/utils.ts (hunk 1 above)
Comment (yorhodes) [reply typescript/infra/src/utils/utils.ts:216]: I think you could reuse the existing enum but I digress

File: typescript/infra/src/agents/key-utils.ts [hunk 8]
```diff
@@ -438,26 +443,25 @@ export async function persistValidatorAddressesToLocalArtifacts(
...
   }
-  // Resolve the relative path
-  const filePath = path.resolve(__dirname, '../../config/aw-multisig.json');
   // Write the updated object back to the file
-  fs.writeFileSync(filePath, JSON.stringify(awMultisigAddresses, null, 2));
+  const dir = path.join(__dirname, '../../config');
+  writeJSON(
+    dir,
+    'aw-multisig.json',
+    JSON.stringify(awMultisigAddresses, null, 2),
```
Comment (yorhodes) [inline typescript/infra/src/agents/key-utils.ts:451]: did you test this?
you don't need to stringify with these functions

Comment (yorhodes) [inline typescript/infra/src/agents/key-utils.ts:451]:
```suggestion
    awMultisigAddresses
```

File: typescript/infra/src/agents/key-utils.ts [hunk 9]
```diff
@@ -438,26 +443,25 @@ export async function persistValidatorAddressesToLocalArtifacts(
...
-  environment: DeployEnvironment,
-  context: Contexts,
-): Address {
-  // Resolve the relative path
-  const filePath = path.resolve(__dirname, `../../config/${role}.json`);
-  const data = fs.readFileSync(filePath, 'utf8');
-  const addresses: LocalRoleAddresses = JSON.parse(data);
+export function fetchLocalKeyAddresses(role: Role): LocalRoleAddresses {
+  try {
+    const dir = path.join(__dirname, '../../config');
```
Comment (yorhodes) [inline typescript/infra/src/agents/key-utils.ts:457]: can you pull this out into a global const?

File: typescript/infra/src/utils/gcloud.ts [hunk 10]
```diff
@@ -31,6 +31,7 @@ export async function fetchGCPSecret(
     );
     output = envVarOverride;
   } else {
+    console.log('FETCHING GCP SECRET: ' + secretName);
```
Comment (yorhodes) [inline typescript/infra/src/utils/gcloud.ts:34]: remove this?

**Review by yorhodes — APPROVED** [review]

File: typescript/infra/src/agents/key-utils.ts (hunk 8 above)
Comment (aroralanuk) [reply typescript/infra/src/agents/key-utils.ts:451]: fixed

File: typescript/infra/src/agents/key-utils.ts (hunk 9 above)
Comment (aroralanuk) [reply typescript/infra/src/agents/key-utils.ts:457]: fixed

**Pull request merged**